contribs/gmf/src/directives/timeslider.js

```diff
@@ -0,0 +1,284 @@
+ * @private
+ * @return {Array<number>}  - List of timestamp representing possible values
+ */
+gmf.TimeSliderController.prototype.getTimeValueList_ = function() {
+  var wmsTime = this.time;
+  var timeValueList = null;
+  var minDate = new Date(this.minValue);
+  var maxDate = new Date(this.maxValue);
+
+  if (wmsTime.values) {
+    timeValueList = [];
+    wmsTime.values.forEach(function(date) {
+      timeValueList.push(new Date(date).getTime());
+    });
+  } else {
+    var maxNbValues = 1024;
+    var endDate = new Date(minDate.getTime());
+    endDate.setFullYear(minDate.getFullYear() + maxNbValues * wmsTime.interval[0]);
+    endDate.setMonth(minDate.getMonth() + maxNbValues * wmsTime.interval[1],
+      minDate.getDate() + maxNbValues * wmsTime.interval[2]);
+    endDate.setSeconds(minDate.getSeconds() + maxNbValues * wmsTime.interval[3]);
+
+    if (endDate > maxDate) {
+      // Transform interval to a list of values when the number
+      // of values is below a threshold (maxNbValues)
+      timeValueList = [];
+      for (var i = 0; ; i++) {
+        var nextDate = new Date(minDate.getTime());
+        nextDate.setFullYear(minDate.getFullYear() + i * wmsTime.interval[0]);
+        nextDate.setMonth(minDate.getMonth() + i * wmsTime.interval[1],
+          minDate.getDate() + i * wmsTime.interval[2]);
+        nextDate.setSeconds(minDate.getSeconds() + i * wmsTime.interval[3]);
+        if (nextDate <= maxDate) {
+          timeValueList.push(nextDate.getTime());
+        } else {
+          break;
+        }
+      }
+    }
+  }
+  return timeValueList;
+};
+
+
+/**
+ * Compute the closest available date from the given timestamp
+ * @param  {number} timestamp selected datetime (in ms format)
+ * @return {number} the closest available datetime (in ms format) from the timestamp
+ * @private
+ */
+gmf.TimeSliderController.prototype.getClosestValue_ = function(timestamp) {
+  if (timestamp <= this.minValue) {
+    return this.minValue;
+  }
+
+  if (timestamp >= this.maxValue) {
+    return this.maxValue;
+  }
+
+  if (this.timeValueList) {
+    // Time stops are defined as a list of values
+    var idx;
+    var lIdx = 0;
+    var rIdx = this.timeValueList.length - 1;
+
+    while ((rIdx - lIdx) > 1) {
+      idx = Math.floor((lIdx + rIdx) / 2);
+      if (this.timeValueList[idx] >= timestamp) {
+        rIdx = idx;
+      } else {
+        lIdx = idx;
+      }
+    }
+
+    var lDist = Math.abs(this.timeValueList[lIdx] - timestamp);
+    var rDist = Math.abs(this.timeValueList[rIdx] - timestamp);
+
+    return this.timeValueList[lDist < rDist ? lIdx : rIdx];
+  } else {
+    // Time stops are defined by a start date plus an interval
+    var targetDate = new Date(timestamp);
+    var startDate = new Date(this.minValue);
+    var bestDate = new Date(this.minValue);
+    var maxDate = new Date(this.maxValue);
+    var bestDistance = Math.abs(targetDate - bestDate);
+
+    for (var i = 1; ; i++) {
```

Comment:
personally I prefer the `for` instance of the `while` and an incremented variable ...